pkg/dl/dl.go

```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
 	}
 }
 
+func useLibDL(action func() error) error {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	return action()
+}
+
+type DLError struct {
+	err string
+}
+
+func (e DLError) Error() string {
+	return e.err
+}
+
+func dlError() error {
+	lastErr := C.dlerror()
+	if lastErr == nil {
+		return nil
+	}
+	return DLError{
+		err: C.GoString(lastErr),
+	}
+}
+
 func (dl *DynamicLibrary) Open() error {
 	name := C.CString(dl.Name)
 	defer C.free(unsafe.Pointer(name))
 
-	handle := C.dlopen(name, C.int(dl.Flags))
-	if handle == C.NULL {
-		return fmt.Errorf("%s", C.GoString(C.dlerror()))
+	var handle unsafe.Pointer
+	if err := useLibDL(func() error {
+		handle = C.dlopen(name, C.int(dl.Flags))
+		if handle == nil {
+			return dlError()
+		}
+		return nil
+	}); err != nil {
+		return err
 	}
+
 	dl.handle = handle
```

Comment:
Is setting (and clearing) the handle not also something that should happen under the lock?

Reply:
Moved under lock